File: APP/android/app/src/main/java/com/airhealth/app/HealthExportState.kt
package com.airhealth.app

enum class HealthExportPlatform(
    val wireValue: String,
    val title: String,
) {
    APPLE_HEALTH("apple_health", "Apple Health"),
    HEALTH_CONNECT("health_connect", "Health Connect"),
}

enum class ExportAuditStatus(
    val wireValue: String,
) {
    SUCCEEDED("succeeded"),
    FAILED("failed"),
}

enum class HealthExportPermissionState(
    val wireValue: String,
) {
    UNKNOWN("unknown"),
    GRANTED("granted"),
    DENIED("denied"),
}

data class CompletedSummaryExportPayload(
    val platform: HealthExportPlatform,
    val sessionId: String,
    val feature: String,
    val resultToken: String,
    val recordedAtEpochMillis: Long,
    val summaryTitle: String,
    val summaryDetail: String,
) {
    fun asWireMap(): Map<String, String> {
        return linkedMapOf(
            "platform" to platform.wireValue,
            "session_id" to sessionId,
            "feature" to feature,
            "result_token" to resultToken,
            "recorded_at_epoch_millis" to recordedAtEpochMillis.toString(),
            "summary_title" to summaryTitle,
            "summary_detail" to summaryDetail,
        )
    }
}

data class ExportAuditRecord(
    val auditId: String,
    val feature: FeatureKind,
    val sessionId: String,
    val platform: HealthExportPlatform,
    val status: ExportAuditStatus,
    val recordedAtEpochMillis: Long,
    val exportedResultToken: String,
    val failureReason: String?,
) {
    val statusLabel: String
        get() = when (status) {
            ExportAuditStatus.SUCCEEDED -> "Exported"
            ExportAuditStatus.FAILED -> "Export failed"
        }
}

data class ExportAuditSurfaceState(
    val permissionState: HealthExportPermissionState,
    val latestStatusLabel: String?,
    val latestPlatformTitle: String?,
    val latestFailureReason: String?,
    val records: List<ExportAuditRecord>,
)

data class HealthExportPermissionStoreState(
    val byPlatform: Map<HealthExportPlatform, HealthExportPermissionState> = emptyMap(),
) {
    fun permissionFor(platform: HealthExportPlatform): HealthExportPermissionState {
        return byPlatform[platform] ?: HealthExportPermissionState.UNKNOWN
    }

    fun setPermission(
        platform: HealthExportPlatform,
        permissionState: HealthExportPermissionState,
    ): HealthExportPermissionStoreState {
        return copy(byPlatform = byPlatform + (platform to permissionState))
    }
}

data class ExportAuditStoreState(
    val records: List<ExportAuditRecord> = emptyList(),
) {
    fun append(record: ExportAuditRecord): ExportAuditStoreState {
        return copy(records = (listOf(record) + records).sortedByDescending { it.recordedAtEpochMillis })
    }

    fun latestRecordFor(feature: FeatureKind): ExportAuditRecord? {
        return records.firstOrNull { it.feature == feature }
    }

    fun surfaceFor(
        feature: FeatureKind,
        permissionState: HealthExportPermissionState,
    ): ExportAuditSurfaceState {
        val featureRecords = records.filter { it.feature == feature }.sortedByDescending { it.recordedAtEpochMillis }
        val latest = featureRecords.firstOrNull()
        return ExportAuditSurfaceState(
            permissionState = permissionState,
            latestStatusLabel = latest?.statusLabel,
            latestPlatformTitle = latest?.platform?.title,
            latestFailureReason = latest?.failureReason,
            records = featureRecords,
        )
    }
}

object HealthExportAdapter {
    fun payloadFor(
        record: PersistedSessionSummaryRecord,
        platform: HealthExportPlatform,
    ): CompletedSummaryExportPayload {
        return CompletedSummaryExportPayload(
            platform = platform,
            sessionId = record.sessionId,
            feature = record.feature.routeId,
            resultToken = record.resultToken,
            recordedAtEpochMillis = record.recordedAtEpochMillis,
            summaryTitle = record.summaryTitle,
            summaryDetail = record.summaryDetail,
        )
    }
}

File: APP/ios/AirHealthApp/App/HealthExportState.swift
import Foundation

enum HealthExportPlatform: String {
    case appleHealth = "apple_health"
    case healthConnect = "health_connect"

    var title: String {
        switch self {
        case .appleHealth:
            return "Apple Health"
        case .healthConnect:
            return "Health Connect"
        }
    }
}

enum ExportAuditStatus: String {
    case succeeded = "succeeded"
    case failed = "failed"
}

enum HealthExportPermissionState: String {
    case unknown = "unknown"
    case granted = "granted"
    case denied = "denied"
}

struct CompletedSummaryExportPayload {
    let platform: HealthExportPlatform
    let sessionID: String
    let feature: String
    let resultToken: String
    let recordedAtEpochMillis: Int64
    let summaryTitle: String
    let summaryDetail: String

    var wireMap: [String: String] {
        [
            "platform": platform.rawValue,
            "session_id": sessionID,
            "feature": feature,
            "result_token": resultToken,
            "recorded_at_epoch_millis": String(recordedAtEpochMillis),
            "summary_title": summaryTitle,
            "summary_detail": summaryDetail,
        ]
    }
}

struct ExportAuditRecord {
    let auditID: String
    let feature: FeatureKind
    let sessionID: String
    let platform: HealthExportPlatform
    let status: ExportAuditStatus
    let recordedAtEpochMillis: Int64
    let exportedResultToken: String
    let failureReason: String?

    var statusLabel: String {
        switch status {
        case .succeeded:
            return "Exported"
        case .failed:
            return "Export failed"
        }
    }
}

struct ExportAuditSurfaceState {
    let permissionState: HealthExportPermissionState
    let latestStatusLabel: String?
    let latestPlatformTitle: String?
    let latestFailureReason: String?
    let records: [ExportAuditRecord]
}

struct HealthExportPermissionStoreState {
    let byPlatform: [HealthExportPlatform: HealthExportPermissionState]

    init(byPlatform: [HealthExportPlatform: HealthExportPermissionState] = [:]) {
        self.byPlatform = byPlatform
    }

    func permission(for platform: HealthExportPlatform) -> HealthExportPermissionState {
        byPlatform[platform] ?? .unknown
    }

    func setPermission(
        _ permissionState: HealthExportPermissionState,
        for platform: HealthExportPlatform
    ) -> HealthExportPermissionStoreState {
        HealthExportPermissionStoreState(byPlatform: byPlatform.merging([platform: permissionState]) { _, new in new })
    }
}

struct ExportAuditStoreState {
    let records: [ExportAuditRecord]

    init(records: [ExportAuditRecord] = []) {
        self.records = records.sorted { $0.recordedAtEpochMillis > $1.recordedAtEpochMillis }
    }

    func append(_ record: ExportAuditRecord) -> ExportAuditStoreState {
        ExportAuditStoreState(records: [record] + records)
    }

    func latestRecord(for feature: FeatureKind) -> ExportAuditRecord? {
        records.first { $0.feature == feature }
    }

    func surface(
        for feature: FeatureKind,
        permissionState: HealthExportPermissionState
    ) -> ExportAuditSurfaceState {
        let featureRecords = records.filter { $0.feature == feature }.sorted { $0.recordedAtEpochMillis > $1.recordedAtEpochMillis }
        let latest = featureRecords.first
        return ExportAuditSurfaceState(
            permissionState: permissionState,
            latestStatusLabel: latest?.statusLabel,
            latestPlatformTitle: latest?.platform.title,
            latestFailureReason: latest?.failureReason,
            records: featureRecords
        )
    }
}

enum HealthExportAdapter {
    static func payloadFor(
        record: PersistedSessionSummaryRecord,
        platform: HealthExportPlatform
    ) -> CompletedSummaryExportPayload {
        CompletedSummaryExportPayload(
            platform: platform,
            sessionID: record.sessionID,
            feature: record.feature.rawValue,
            resultToken: record.resultToken,
            recordedAtEpochMillis: record.recordedAtEpochMillis,
            summaryTitle: record.summaryTitle,
            summaryDetail: record.summaryDetail
        )
    }
}
